### ISSUE TITLE
clarify what functions can be called on uncommitted datatypes

### ISSUE THREAD

**Issue by jeffhammond** [mpi-forum/mpi-issues#17]
This text
```
4.1.9 Commit and Free
A datatype object has to be committed before it can be used in a communication. 
As an argument in datatype constructors, uncommitted and also committed datatypes 
can be used. There is no need to commit basic datatypes. They are “pre-committed.”
```
leaves some ambiguity since the functions `MPI_TYPE_SIZE(_X)` and `MPI_TYPE_GET_{CONTENTS,ENVELOPE,EXTENT(_X),TRUE_EXTENT(_X)}` lie outside of the union of "communication" and "datatype constructors".

It would be helpful to users if the text was unambiguous about the functions that can and cannot be called on uncommitted datatypes.

**Comment by bosilca** [comment]
The text clearly states what the major constraint is: datatype must be committed before being used in a communication. All other cases do not require the commit stage. However, I agree with you that the second part is confusing, there is no need to over-clarify that a non-committed datatype can be used as an argument to datatype constructors, or if we do we should define it as an example.

**Comment by jeffhammond** [comment]
@bosilca My goal is to clarify the case of functions that are _neither_ communication functions nor datatype constructors.  It may be obvious to the Forum, but I believe a power user was confused about this, which is why I created the original issue.

**Comment by bosilca** [comment]
@jeffhammond we are over-specifying. The text should be simple: only datatypes used in communications should be committed, all other usages do not require a committed datatype.

**Comment by jeffhammond** [comment]
@bosilca It sounds like we should replace "datatype constructors" with "other functions" or "non-communication MPI functions".

**Comment by bosilca** [comment]
@jeffhammond we should be careful about how we define "non-communication functions". Are I/O included?

**Comment by jeffhammond** [comment]
@bosilca Hence my original proposal: _It would be helpful to users if the text was unambiguous about the functions that can and cannot be called on uncommitted datatypes._

**Comment by thender** [comment]
Should uncommitted datatypes be freed via MPI_TYPE_FREE()? My use-case is nested datatypes where "leaf" datatypes need not be committed (per 4.1.9).

**Comment by wgropp** [comment]
Yes, they will eventually need to be freed.  They can be freed once the type is no longer needed explicitly, even if they are used within another type.  In fact, in many cases, this is the correct approach - create the datatypes, free all but the ones that are committed and used in subsequent communication.

Bill

William Gropp
Director, Parallel Computing Institute
Thomas M. Siebel Chair in Computer Science
Chief Scientist, NCSA
University of Illinois Urbana-Champaign

**Comment by jeffhammond** [comment]
Below is a list of functions (not complete because of the sloppy way I parsed MPICH's `mpi.h`) that take a datatype argument and whether or not I think the datatype needs to be committed.

Summary:
- Yes for all communication functions (point-to-point, collectives, RMA), although `MPI_Compare_and_swap` and `MPI_Fetch_and_op` only accept pre-committed types anyways.
- Yes for all file I/O functions, although we could make exceptions for `MPI_File_get_type_extent`, `MPI_File_set_view` and `MPI_File_get_view`.
- Yes for `MPI_Datarep_conversion_function` and `MPI_Datarep_extent_function`, which imply `MPI_Register_datarep`, although I only learned about these functions today and don't care what we decide.
- Yes for `MPI_Pack`, `MPI_Unpack`, `MPI_Pack_external`, `MPI_Unpack_external` and `MPI_Reduce_local` because they manipulate buffer data like communication functions.
- No for all MPI datatype functions.
- No for `MPI_Status_set_elements`, `MPI_Status_set_elements_x`, `MPI_Get_count`, `MPI_Get_elements`, `MPI_Get_elements_x`, `MPI_Pack_external_size` and `MPI_Pack_size`.

```c
MPI_Send - yes
MPI_Recv - yes
MPI_Get_count - no
MPI_Bsend - yes
MPI_Ssend - yes
MPI_Rsend - yes
MPI_Isend - yes
MPI_Ibsend - yes
MPI_Issend - yes
MPI_Irsend - yes
MPI_Irecv - yes
MPI_Send_init - yes
MPI_Bsend_init - yes
MPI_Ssend_init - yes
MPI_Rsend_init - yes
MPI_Recv_init - yes
MPI_Sendrecv - yes
MPI_Sendrecv_replace - yes
MPI_Type_contiguous - no
MPI_Type_vector - no
MPI_Type_hvector - no
MPI_Type_extent - no
MPI_Type_size - no
MPI_Type_lb - no
MPI_Type_ub - no
MPI_Type_commit - N/A (already covered)
MPI_Type_free - no
MPI_Get_elements - no
MPI_Pack - yes
MPI_Pack_size - no
MPI_Bcast - yes
MPI_Gather - yes
MPI_Gatherv - yes
MPI_Scatter - yes
MPI_Allgather - yes
MPI_Allgatherv - yes
MPI_Alltoall - yes
MPI_Exscan - yes
MPI_Reduce - yes
MPI_Allreduce - yes
MPI_Scan - yes
MPI_Accumulate - yes
MPI_Get - yes
MPI_Get_accumulate - yes
MPI_Put - yes
MPI_Raccumulate - yes
MPI_Rget_accumulate - yes
MPI_Rget - yes
MPI_Rput - yes
MPI_Status_set_elements - no
MPI_Type_delete_attr - no
MPI_Type_dup - no
MPI_Type_get_attr - no
MPI_Type_get_contents - no
MPI_Type_get_envelope - no
MPI_Type_get_name - no
MPI_Type_set_attr - no
MPI_Type_set_name - no
MPI_Type_match_size - no
MPI_Pack_external_size - no
MPI_Type_create_hvector - no
MPI_Type_create_resized - no
MPI_Type_get_extent - no
MPI_Type_get_true_extent - no
MPI_Type_create_f90_integer - no
MPI_Type_create_f90_real - no
MPI_Type_create_f90_complex - no
MPI_Reduce_local - yes
MPI_Imrecv - yes
MPI_Mrecv - yes
MPI_Ibcast - yes
MPI_Igather - yes
MPI_Igatherv - yes
MPI_Iscatter - yes
MPI_Iallgather - yes
MPI_Iallgatherv - yes
MPI_Ialltoall - yes
MPI_Ireduce - yes
MPI_Iallreduce - yes
MPI_Iscan - yes
MPI_Iexscan - yes
MPI_Ineighbor_allgather - yes
MPI_Ineighbor_allgatherv - yes
MPI_Ineighbor_alltoall - yes
MPI_Neighbor_allgather - yes
MPI_Neighbor_allgatherv - yes
MPI_Neighbor_alltoall - yes
MPI_Get_elements_x - no
MPI_Status_set_elements_x - no
MPI_Type_get_extent_x - no
MPI_Type_get_true_extent_x - no
MPI_Type_size_x - no
MPI_File_set_view - yes
MPI_File_get_view - yes
MPI_File_read_at - yes
MPI_File_iread_at - yes
MPI_File_read - yes
MPI_File_read_all - yes
MPI_File_write - yes
MPI_File_write_all
MPI_File_iread - yes
MPI_File_iwrite - yes
MPI_File_read_shared - yes
MPI_File_write_shared - yes
MPI_File_iread_shared - yes
MPI_File_iwrite_shared - yes
MPI_File_read_ordered - yes
MPI_File_write_ordered - yes
MPI_File_read_all_begin - yes
MPI_File_write_all_begin - yes
MPI_File_read_ordered_begin - yes
MPI_File_write_ordered_begin - yes
MPI_File_get_type_extent - no
MPI_File_iread_all - yes
MPI_File_iwrite_all - yes
```

**Comment by wesbland** [comment]
I’m going to propose moving this to MPI 5.0. There’s more discussion to be had here. If someone objects and thinks we’ll be ready to read this soon, leave a comment and we can discuss bringing it back into MPI 4.1.
